### DISCUSSION TITLE
Danke und Frage zu Modbus Zähler

### DISCUSSION THREAD

**Discussion by lfischer85** [camueller/SmartApplianceEnabler#494] [Installation und Betrieb]
Hallo Axel

Erst mal vielen Dank für diese tolle Software und die ausführliche Dokumentation. Die Installation der Version 2.2.1 und die erste Einrichtung eines Gerätes mit Modbus Zähler hat soweit geklappt, nun stehe ich aber an:
Die Logs zeigen das Auslesen des Modbus Zählers mit plausiblen Werten, anschliessend aber wird daraus eine Leistung von "0" berechnet. Diese wird dann wiederum korrekt im Sunny Portal angezeigt zu meinem Gerät.

>2023-03-08 08:39:35,152 TRACE [Timer-0] d.a.s.m.e.ReadInputRegisterExecutor [ReadInputRegisterExecutor.java:42] F-19851191-010003000001-00: Reading input register=145 requestWords=2
>
>2023-03-08 08:39:35,377 DEBUG [Timer-0] d.a.s.m.e.ReadInputRegisterExecutor [ReadInputRegisterExecutor.java:56] F-19851191-010003000001-00: Input register=145 value=[97, 32553]
>
>2023-03-08 08:39:35,377 DEBUG [Timer-0] d.a.s.m.t.FloatValueTransformer [IntegerValueTransformer.java:43] F-19851191-010003000001-00: transformed value=6389545
>
>2023-03-08 08:39:35,379 DEBUG [Timer-0] d.a.s.m.PollEnergyMeter [PollEnergyMeter.java:105] F-19851191-010003000001-00: Calculating power from energy: power=0 currentEnergyCounter=0 previousEnergyCounter=0 diffEnergy=0 diffTime=59999

Log Zeile 3 oben zeigt den plausiblen Wert für den Zählerstand, der zählt auch alle paar Sekunden hoch. Die Last beträgt ca. 55 Watt.
In Log Zeile 4 steht dann "Calculating power from energy: power=0".

Was mache ich noch falsch?

Vielen Dank fürs Lesen und freundliche Grüsse
Lukas

**Comment by camueller** [comment]
Der Zählerstand muss in kWh sein, da scheint "6389545" nicht zu passen. Ggf. musst Du einen Umrechnungsfaktor angeben, aber die meisten Modbus-Zähler liefern dafür eine Gleitkommazahl. Vielleicht liest Du ein falsches Register aus.

**Reply by lfischer85** [reply]
Danke für die rasche Antwort :) 
Der "Zähler" ist eine APC USV, das Register gibt die Verbrauchte Energie in Wattstunden aus als "UINT32", 2 Datenworte breit. Die Zahl passt also so, der Umrechnungsfaktor ist entsprechend als 0.001 eingetragen um auf kWh zu kommen.

Trotzdem scheint da etwas schief zu gehen bei der Verarbeitung der Daten.

Edit: Die USV meldet im Webinterface "Output Energy Usage: 6389.72 kWh", das kommt also hin.

**Comment by camueller** [comment]
Was hast Du denn überhapt konfiguriert außer dem Zähler? Poste doch mal einen Log-Auszug ...

**Comment by lfischer85** [comment]
Konfiguriert habe ich ein Gerät mit
ID: *F-19851191-010003000001-00*
Hersteller: *APC*
Bezeichnung: *USV01*
Typ: *Sonstige*
Seriennummer: *1*
Min. Leistunsaufnahme: *(leer)*
Max. Leistungsaufnahme: *100W*
Unterbrechungen erlaubt: *ausgeschaltet*

Schalter zu USV01:
Typ: *Immer eingeschaltet*

Anbei der Log Auszug:
[LogAuszug SAE.txt](https://github.com/camueller/SmartApplianceEnabler/files/10922161/LogAuszug.SAE.txt)

**Comment by camueller** [comment]
Du hast immer noch keine korrekten Zählerwerte:
```
transformed value=6389994
```

**Comment by lfischer85** [comment]
Der Zähler ist wie folgt konfiguriert:
Typ: *Modbus*
Modbus: *USV01*
Slave-Adresse: *1*
Register-Adresse: *145*
Register-Typ: *Input (FC=4)*
Wert-Typ: *Integer (32bit)*
Datenwörter: *2*
Byte Reihenfolge: *Big Endian (Standard)*
Umrechnungsfaktor: *0.001*
Parameter: *Zählerstand*
Regex für Extraktion: *(leer)*

Das transformed Value entspricht genau dem Registerinhalt. Gemäss Dokumentation des Zählers sind dies Wattstunden, der Wert deckt sich mit den Angaben im Webinterface.

Müsste das transformed value den Umrechnungsfaktor mit berücksichtigen? Dann klappt da etwas nicht.

**Comment by camueller** [comment]
Du musst "Integer als Float" einstellten als Wert-Typ.

**Comment by lfischer85** [comment]
Autsch, das war nun wirklich zu einfach.. Herzlichen Dank für Deine Unterstützung! Nun klappt es wie erwartet.

`2023-03-08 17:02:49,305 DEBUG [Timer-0] d.a.s.m.t.Integer2FloatValueTransformer [Integer2FloatValueTransformer.java:47] F-19851191-010003000001-00: transformed value=6390.032
2023-03-08 17:02:49,308 DEBUG [Timer-0] d.a.s.m.PollEnergyMeter [PollEnergyMeter.java:105] F-19851191-010003000001-00: Calculating power from energy: power=60 currentEnergyCounter=6390.032 previousEnergyCounter=6390.031 diffEnergy=0.001 diffTime=60000
`

**Reply by camueller** [reply]
3 Nachkommstellen ist nicht genau genug bei kleiner Leistungaufnahme!

**Reply by lfischer85** [reply]
Danke für den Hinweis! Das stimmt, ich habe so Abweichungen zwischen 5 und 11%. Das ist aber in diesem Fall für mich verschmerzbar.

**Reply by Biker-Lab** [reply]
> 3 Nachkommstellen ist nicht genau genug bei kleiner Leistungaufnahme!

Hallo Axel,  ich hänge mich mal hier dran : Kann man von einem SDM72DM-V2 mehr als 3 Nachkommastellen per Modbus auslesen?
Ich habe meinen SDM72 jetzt von S0 auf Modbus umgestellt. Konfiguration wie im Beispielbild.

Auslesen klappt, im Log unter ModBusElectricityMeter : _transformed value=61.893001556396484_  _Current energy meter value: 61.893001556396484kWh_ 
Wenn aber dann die Leistung aus der Energie berechnet wird sind es nur 3 Nachkommastellen : 
_Calculating power from energy: power=0 currentEnergyCounter=61.895 previousEnergyCounter=61.895 diffEnergy=0 diffTime=60000_

Das ist zu ungenau, teilweise kommt dann 0W dabei raus.
Ich habe schon mehrere Optionen (Werttyp, Datenwörter, Byte Reihenfolge) probiert, aber da kommen dann ganz andere Zahlen zustande.
Danke

**Reply by camueller** [reply]
Die Leistung wird als Ganzzahlwert an den SHM übergeben. Aber die Zähelrstandsdifferenzen, aus denen die Leistung berechnet wird, sollten mindestens 5 Nachkomstellen haben. Warum es bei Dir nur 3 sind, kann ich nicht sagen. Bei meinem SDM72DM sieht die Zeile so aus:
```
2023-06-14 07:57:00,104 DEBUG [Timer-0] d.a.s.m.PollEnergyMeter [PollEnergyMeter.java:108] F-00000001-000000000019-00: Calculating power from energy: power=0 currentEnergyCounter=4897.91895 previousEnergyCounter=4897.91895 diffEnergy=0 diffTime=60000
```
Verwendest Du die aktuelle Version des SAE?

Hier meine Konfiguration:
<img width="778" alt="Screenshot_20230621_074544" src="https://github.com/camueller/SmartApplianceEnabler/assets/4489628/3d5244f9-cb9a-46ac-b92a-9e9669e3c16d">

**Reply by Biker-Lab** [reply]
> Verwendest Du die aktuelle Version des SAE?

Hatte noch 2.2.2 installiert und jetzt Update auf 2.3.0 gemacht, aber da ändert sich nichts.

Die Konfiguration habe ich genauso eingestellt.

Sind die vielen Nachkommastellen bei transformed value und current energy meter value denn normal? Sehr oft ist dort die 4. und 5. Nachkommastelle 0

_2023-06-21` 17:56:35,941 DEBUG [MQTT Call: F-06121973-000000000001-00-RuntimeRequest-58] d.a.s.m.ModbusElectricityMeter [ModbusElectricityMeter.java:208] F-06121973-000000000001-00: Current energy meter value: 67.50800323486328kWh
2023-06-21 17:56:35,939 DEBUG [MQTT Call: F-06121973-000000000001-00-RuntimeRequest-58] d.a.s.m.t.FloatValueTransformer [FloatValueTransformer.java:44] F-06121973-000000000001-00: transformed value=67.50800323486328
2023-06-21 17:56:35,938 DEBUG [MQTT Call: F-06121973-000000000001-00-RuntimeRequest-58] d.a.s.m.e.ReadInputRegisterExecutor [ReadInputRegisterExecutor.java:56] F-06121973-000000000001-00: Input register=342 value=[17031, 1049]_

**Reply by camueller** [reply]
Ich kann nicht nachvollziehen, auf welche Zeile Du Dich beziehst - bitte angeben.

**Reply by Biker-Lab** [reply]
Ich meinte, ob diese vielen Nachkommastellen in fett normal sind oder ob hier schon nur 5 Nachkommastellen erscheinen müssen.

2023-06-21` 17:56:35,941 DEBUG [MQTT Call: F-06121973-000000000001-00-RuntimeRequest-58] d.a.s.m.ModbusElectricityMeter [ModbusElectricityMeter.java:208] F-06121973-000000000001-00: **Current energy meter value: 67.50800323486328kWh**
2023-06-21 17:56:35,939 DEBUG [MQTT Call: F-06121973-000000000001-00-RuntimeRequest-58] d.a.s.m.t.FloatValueTransformer [FloatValueTransformer.java:44] F-06121973-000000000001-00: **transformed value=67.50800323486328**
2023-06-21 17:56:35,938 DEBUG [MQTT Call: F-06121973-000000000001-00-RuntimeRequest-58] d.a.s.m.e.ReadInputRegisterExecutor [ReadInputRegisterExecutor.java:56] F-06121973-000000000001-00: Input register=342 value=[17031, 1049]

**Comment by camueller** [comment]
Jede Zahl hat einen Typ, welcher die Anzahl der Ziffern und damit die Genauigkeit festlegt. Was Du im SAE-Log siehst, ist manchmal die Zahl mit voller Genauigkeit und manchmal formatiert auf eine bestimme Anzahl von Nachkommstellen. Es geht nicht darum, irgendwo genau 5 Nachkommstellen zu haben, sondern **mindestens** 5 Nachkommastellen, weil ansonsten aus der Differenz solcher Zählerstände keine Leistung mit niedrigem Wert berechnet werden kann.
